File: iosApp/iosApp/AdsConsentManager.swift
import GoogleMobileAds
import common

/// iOS composition root: wires the UMP-backed `UmpConsentGateway` to the shared `ConsentCoordinator`,
/// then starts the Mobile Ads SDK once ads may be requested.
enum AdsConsentManager {
    // Retain the coordinator for the duration of the async consent flow.
    private static var coordinator: ConsentCoordinator?

    /// Call once after the UI is on screen (the consent form is presented over the top view controller).
    static func gatherConsentThenStartAds() {
        let coordinator = ConsentCoordinator(
            gateway: UmpConsentGateway(),
            onAdsReady: { GADMobileAds.sharedInstance().start(completionHandler: nil) }
        )
        self.coordinator = coordinator
        coordinator.gatherConsentThenInitializeAds()
    }
}

File: androidApp/src/main/kotlin/com/oztechan/adtrack/android/ads/AdsConsentManager.kt
/*
 * Copyright (c) 2026 Mustafa Ozhan. All rights reserved.
 */

package com.oztechan.adtrack.android.ads

import android.app.Activity
import com.google.android.gms.ads.MobileAds
import com.oztechan.adtrack.ads.consent.ConsentCoordinator
import kotlin.concurrent.thread

/**
 * Wires the UMP-backed [UmpConsentGateway] to the shared [ConsentCoordinator] for this Activity:
 * gather consent, then initialize the Mobile Ads SDK once ads may be requested.
 */
class AdsConsentManager(activity: Activity) {

    private val coordinator = ConsentCoordinator(
        gateway = UmpConsentGateway(activity),
        onAdsReady = {
            // Google recommends initializing the Mobile Ads SDK off the main thread, as it does I/O.
            thread { MobileAds.initialize(activity.applicationContext) }
        }
    )

    fun gatherConsentThenInitializeAds() = coordinator.gatherConsentThenInitializeAds()
}
